src/libraries/System.Private.CoreLib/src/System/IO/Strategies/BufferedFileStreamStrategy.cs

```diff
@@ -649,19 +649,19 @@ public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationTo
                 bool releaseTheLock = true;
                 try
                 {
-                    // hot path #1 if the write completely fits into the buffer, we can complete synchronously:
-                    if (_bufferSize - _writePos >= buffer.Length)
+                    // hot path #1: the write buffer is empty and buffering would not be beneficial
+                    if (_writePos == 0 && buffer.Length >= _bufferSize)
                     {
+                        return _strategy.WriteAsync(buffer, cancellationToken);
```

Comment:
This path calls to WriteAsync and directly returns its ValueTask, with the underlying operation potentially / likely still in flight, and then we immediately exit the semaphore in the finally block, which means a subsequent call could successfully acquire the semaphore even while this async operation is still in flight.  Is that intended?

Reply:
>  Is that intended?

Yes. The idea here is to bypass the buffering logic if nothing has been read|written to buffer yet and the buffering would not be beneficial (user buffer size >= _bufferSize). It basically tries to behave as a `FileStream` created with `bufferSize=1` (no serialization of the async operations)

Reply:
Should this be awaited then? It looks like `WriteAsyncSlowPath` always awaits for WriteAsync calls before releasing the semaphore.

Reply:
> It basically tries to behave as a FileStream created with bufferSize=1 (no serialization of the async operations)

Ok

Reply:
I've added a comment that explains why we are doing it and why it's possible to do it that way.